### PR TITLE
Tracing

### PR DESCRIPTION
**Pull request by irevoire**
# Pull Request

- [ ] Hide the parameters of the process batch
- [x] Make actix-web trace every call on every route
- [x] Remove all `env_logger`/`logs` dependencies
- [x] Be able to enable or disable the memory measurement using the `/logs` route parameters

See the following product discussion: https://github.com/orgs/meilisearch/discussions/721

Supersedes https://github.com/meilisearch/meilisearch/pull/4338

## Related issue
Fixes https://github.com/meilisearch/meilisearch/issues/4317

## What does this PR do?

Update the format of the logs from:
```
[2024-02-06T14:54:11Z INFO  actix_server::builder] starting 10 workers
```

to

```
2024-02-06T13:58:14.710803Z  INFO actix_server::builder: 200: starting 10 workers
```

First, run meilisearch with the route enabled via the feature flag:
- `cargo run --experimental-enable-logs-route`
- Or at runtime by sending the following payload:
```
curl \
  -X PATCH 'http://localhost:7700/experimental-features/' \
  -H 'Content-Type: application/json'  \
--data-binary '{
    "logsRoute": true
  }'
```

Then gather data from meilisearch by calling for example:
```
curl \
	-X POST http://localhost:7700/logs \
	-H 'Content-Type: application/json' \
	--data-binary '{
	    "mode": "fmt",
            "target": "milli=trace"
    }'
```

Once your operation is over, tell meilisearch to stop the route:
```
curl \
	-X DELETE http://localhost:7700/logs
```

----

In the case you’re profiling code, you will be interested by the next command that converts the output of the route to a format that the firefox profiler can understand.

```bash
cargo run --release --bin trace-to-firefox -- 2024-01-17_17:07:55-indexing-trace.json
```

Then go to https://profiler.firefox.com and load it.
Note that we can also share the profiles using the https://share.firefox.dev website.

### CODE REVIEW

File: meilisearch/src/error.rs [hunk 1]
```diff
@@ -12,6 +12,8 @@ pub enum MeilisearchHttpError {
     #[error("A Content-Type header is missing. Accepted values for the Content-Type header are: {}",
             .0.iter().map(|s| format!("`{}`", s)).collect::<Vec<_>>().join(", "))]
     MissingContentType(Vec<String>),
+    #[error("Log route is currently used by someone else.")]
```
Comment (dureuill) [inline meilisearch/src/error.rs:15]:
```suggestion
    #[error("The logs/stream route is currently in use by someone else.")]
```

File: meilisearch/src/lib.rs [hunk 2]
```diff
@@ -123,7 +136,7 @@ pub fn create_app(
             .allow_any_method()
             .max_age(86_400), // 24h
     )
-    .wrap(actix_web::middleware::Logger::default())
+    .wrap(tracing_actix_web::TracingLogger::default())
```
Comment (dureuill) [inline meilisearch/src/lib.rs:139]: the default tracing logger is a bit noisy I think:

```
2024-02-07T14:54:01.245351Z  INFO HTTP request{http.method=GET http.route=/tasks/{task_id} http.flavor=1.1 http.scheme=http http.host=localhost:7700 http.client_ip=::1 http.user_agent=mieli/0.28.2 http.target=/tasks/0 otel.name=HTTP GET /tasks/{task_id} otel.kind="server" request_id=537f2c46-4b07-43f9-bb7c-eac494ba7434}: tracing_actix_web::root_span_builder: 41: new
```

Looks like we can actually customize its output, should we consider doing so?

File: meilisearch/src/option.rs [hunk 3]
```diff
@@ -309,6 +310,13 @@ pub struct Opt {
     #[serde(default)]
     pub experimental_enable_metrics: bool,
 
+    /// Experimental logs route feature. For more information, see: <https://github.com/orgs/meilisearch/discussions/721>
+    ///
+    /// Enables the log route on the `POST /logs` endpoint and the `DELETE /logs` to stop receiving logs.
```
Comment (dureuill) [inline meilisearch/src/option.rs:315]:
```suggestion
    /// Enables the log route on the `POST /logs/stream` endpoint and the `DELETE /logs` to stop receiving logs.
```

File: meilisearch/src/routes/logs.rs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            .route(web::post().to(SeqHandler(get_logs)))
+            .route(web::delete().to(SeqHandler(cancel_logs))),
+    );
+}
+
+#[derive(Debug, Default, Clone, Copy, Deserr, PartialEq, Eq)]
+#[deserr(rename_all = lowercase)]
+pub enum LogMode {
+    #[default]
+    Fmt,
```
Comment (dureuill) [inline meilisearch/src/routes/logs.rs:39]: Maybe for later: "fmt" looks too implementation-oriented, I think the name should reflect the intended purpose of the log.

So here "Human" or "HumanReadable" or "Pretty" or "Print", something like that (in decreasing order of preference)

File: meilisearch/src/routes/logs.rs [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+pub fn configure(cfg: &mut web::ServiceConfig) {
+    cfg.service(
+        web::resource("stream")
+            .route(web::post().to(SeqHandler(get_logs)))
+            .route(web::delete().to(SeqHandler(cancel_logs))),
+    );
+}
+
+#[derive(Debug, Default, Clone, Copy, Deserr, PartialEq, Eq)]
+#[deserr(rename_all = lowercase)]
```
Comment (dureuill) [inline meilisearch/src/routes/logs.rs:36]: don't we generally rename to camelCase?

File: meilisearch/src/routes/logs.rs [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+#[derive(Clone, Debug)]
+struct MyTargets(Targets);
+
+/// Simple wrapper around the `ParseError` from `tracing_subscriber` to implement `MergeWithError` on it.
+#[derive(Debug, thiserror::Error)]
+enum MyParseError {
+    #[error(transparent)]
+    ParseError(#[from] tracing_subscriber::filter::ParseError),
+    #[error(
+        "Empty string is not a valid target. If you want to get no logs use `OFF`. Usage: `info`, `info:meilisearch`, or you can write multiple filters in one target: `index_scheduler=info,milli=trace`"
```
Comment (dureuill) [inline meilisearch/src/routes/logs.rs:53]: does "info:meilisearch" actually works as a syntax? I thought it was "meilisearch=info" (like in the subsequent list example)

File: meilisearch/src/routes/tasks.rs [hunk 7]
```diff
@@ -263,6 +263,7 @@ async fn get_tasks(
     req: HttpRequest,
     analytics: web::Data<dyn Analytics>,
 ) -> Result<HttpResponse, ResponseError> {
+    tracing::info!("You called tasks");
```
Comment (dureuill) [inline meilisearch/src/routes/tasks.rs:266]:
```suggestion
```

File: meilisearch/tests/tasks/webhook.rs [hunk 8]
```diff
@@ -44,10 +44,6 @@ struct WebhookHandle {
 }
 
 async fn create_webhook_server() -> WebhookHandle {
-    let mut log_builder = env_logger::Builder::new();
-    log_builder.parse_filters("info");
-    log_builder.init();
```
Comment (dureuill) [inline meilisearch/tests/tasks/webhook.rs:49]: what was this for? what do we lose by removing it?

File: meilisearch/src/routes/features.rs [hunk 9]
```diff
@@ -33,8 +33,9 @@ async fn get_features(
     let features = index_scheduler.features();
 
     analytics.publish("Experimental features Seen".to_string(), json!(null), Some(&req));
-    debug!("returns: {:?}", features.runtime_features());
-    HttpResponse::Ok().json(features.runtime_features())
+    let features = features.runtime_features();
+    debug_span!("Get features", returns = ?features);
```
Comment (dureuill) [inline meilisearch/src/routes/features.rs:37]:
```suggestion
    tracing::debug!(returns = ?features, "Get features");
```

File: meilisearch/src/error.rs (hunk 1 above)
Comment (dureuill) [reply meilisearch/src/error.rs]: done

File: meilisearch/src/main.rs [hunk 10]
```diff
@@ -9,25 +10,42 @@ use actix_web::HttpServer;
...
-    let log_filters = format!(
-        "{},h2=warn,hyper=warn,tokio_util=warn,tracing=warn,rustls=warn,mio=warn,reqwest=warn",
-        opt.log_level
+fn setup(opt: &Opt) -> anyhow::Result<LogRouteHandle> {
+    let (route_layer, route_layer_handle) = tracing_subscriber::reload::Layer::new(default_layer());
+    let route_layer: tracing_subscriber::reload::Layer<_, _> = route_layer;
+
+    let subscriber = tracing_subscriber::registry().with(route_layer).with(
+        tracing_subscriber::fmt::layer()
+            .with_line_number(true)
```
[the remark this answers is not in the document]
Comment (dureuill) [reply meilisearch/src/main.rs]: done

File: meilisearch/src/option.rs (hunk 3 above)
Comment (dureuill) [reply meilisearch/src/option.rs]: done

File: meilisearch/src/routes/dump.rs [hunk 11]
```diff
@@ -32,6 +32,6 @@ pub async fn create_dump(
     let task: SummarizedTaskView =
         tokio::task::spawn_blocking(move || index_scheduler.register(task)).await??.into();
 
-    debug!("returns: {:?}", task);
+    debug_span!("Create dump", returns = ?task);
```
[the remark this answers is not in the document]
Comment (dureuill) [reply meilisearch/src/routes/dump.rs]: done by tamo

File: meilisearch/src/routes/logs.rs (hunk 4 above)
Comment (irevoire) [reply meilisearch/src/routes/logs.rs:39]: Yeah, tbh I was also a bit against fmt but had no idea oh what to write and was finding `human readable` too long to write.
Didn’t think of `human` though 😅 
That looks good; if you’re ok I would like to do it in this PR so you and Kero don’t have to update your code again once the PR lands

Comment (dureuill) [reply meilisearch/src/routes/logs.rs:39]: The benchmark isn't using the human mode, so go ahead :-)

Comment (dureuill) [reply meilisearch/src/routes/logs.rs:39]: it doesn't really convey the "activity" and is subjective

File: milli/src/update/index_documents/extract/mod.rs [hunk 12]
```diff
@@ -230,7 +231,7 @@ pub(crate) fn data_from_obkv_documents(
         extract_facet_number_docids,
         merge_deladd_cbo_roaring_bitmaps,
         TypedChunk::FieldIdFacetNumberDocids,
-        "field-id-facet-number-docids",
+        "field-id-facet-number-docidsdexing::details, ",
```
[the remark this answers is not in the document]
Comment (dureuill) [reply milli/src/update/index_documents/extract/mod.rs:234]: hey, could be me

[the remark this answers is not in the document]
Comment (irevoire) [reply milli/src/update/index_documents/extract/mod.rs]: It’s probably kefir, he wont stop shoving his nose on my keyboard

File: meilisearch/src/routes/logs.rs (hunk 4 above)
Comment (irevoire) [reply meilisearch/src/routes/logs.rs]: Yeah, I'm more on Louis's side on this one. I went with human

**Review by dureuill — DISMISSED** [review]
Let's usher into a new era of structured logs 🚀 

bors merge

**Review by dureuill — APPROVED** [review]
Of course the CI would fail on `rustfmt` after I approved with such dramatic words

bors merge

**Pull request merged**